File: graphlib-1/go/lib/paths.go
package lib

import (
	"container/heap"
	"math"
	"sort"
)

type pqItem struct {
	node     string
	priority float64
	index    int
}

type priorityQueue []*pqItem

func (pq priorityQueue) Len() int            { return len(pq) }
func (pq priorityQueue) Less(i, j int) bool  { return pq[i].priority < pq[j].priority }
func (pq priorityQueue) Swap(i, j int)       { pq[i], pq[j] = pq[j], pq[i]; pq[i].index = i; pq[j].index = j }
func (pq *priorityQueue) Push(x interface{}) { n := len(*pq); item := x.(*pqItem); item.index = n; *pq = append(*pq, item) }
func (pq *priorityQueue) Pop() interface{}   { old := *pq; n := len(old); item := old[n-1]; old[n-1] = nil; item.index = -1; *pq = old[0 : n-1]; return item }

func (s *GraphStore) ShortestPath(graphID, start, end string) map[string]interface{} {
	s.mu.RLock()
	defer s.mu.RUnlock()
	g, ok := s.graphs[graphID]
	if !ok {
		return map[string]interface{}{"exists": false, "path": []string{}, "distance": float64(-1)}
	}
	if start == end && g.Nodes[start] {
		return map[string]interface{}{"exists": true, "path": []string{start}, "distance": float64(0)}
	}
	if !g.Nodes[start] || !g.Nodes[end] {
		return map[string]interface{}{"exists": false, "path": []string{}, "distance": float64(-1)}
	}
	if !g.Weighted {
		return s.bfsShortestPath(g, start, end)
	}
	return s.dijkstraShortestPath(g, start, end)
}

func (s *GraphStore) bfsShortestPath(g *Graph, start, end string) map[string]interface{} {
	parent := make(map[string]string)
	visited := make(map[string]bool)
	queue := []string{start}
	visited[start] = true
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		if current == end {
			path := []string{}
			for node := end; node != ""; node = parent[node] {
				path = append([]string{node}, path...)
				if node == start {
					break
				}
			}
			return map[string]interface{}{"exists": true, "path": path, "distance": float64(len(path) - 1)}
		}
		neighbors := []string{}
		for neighbor := range g.AdjList[current] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)
		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				visited[neighbor] = true
				parent[neighbor] = current
				queue = append(queue, neighbor)
			}
		}
	}
	return map[string]interface{}{"exists": false, "path": []string{}, "distance": float64(-1)}
}

func (s *GraphStore) dijkstraShortestPath(g *Graph, start, end string) map[string]interface{} {
	dist := make(map[string]float64)
	parent := make(map[string]string)
	for node := range g.Nodes {
		dist[node] = math.Inf(1)
	}
	dist[start] = 0
	pq := make(priorityQueue, 0)
	heap.Init(&pq)
	heap.Push(&pq, &pqItem{node: start, priority: 0})
	for pq.Len() > 0 {
		item := heap.Pop(&pq).(*pqItem)
		current := item.node
		if item.priority > dist[current] {
			continue
		}
		if current == end {
			path := []string{}
			for node := end; node != ""; node = parent[node] {
				path = append([]string{node}, path...)
				if node == start {
					break
				}
			}
			return map[string]interface{}{"exists": true, "path": path, "distance": dist[end]}
		}
		for neighbor, weight := range g.AdjList[current] {
			newDist := dist[current] + weight
			if newDist < dist[neighbor] {
				dist[neighbor] = newDist
				parent[neighbor] = current
				heap.Push(&pq, &pqItem{node: neighbor, priority: newDist})
			}
		}
	}
	return map[string]interface{}{"exists": false, "path": []string{}, "distance": float64(-1)}
}

func (s *GraphStore) AllShortestPaths(graphID, start string) map[string]interface{} {
	s.mu.RLock()
	defer s.mu.RUnlock()
	g, ok := s.graphs[graphID]
	if !ok {
		return map[string]interface{}{"distances": map[string]float64{}, "paths": map[string][]string{}, "unreachable": []string{}}
	}
	if !g.Nodes[start] {
		unreachable := []string{}
		for node := range g.Nodes {
			unreachable = append(unreachable, node)
		}
		sort.Strings(unreachable)
		return map[string]interface{}{"distances": map[string]float64{}, "paths": map[string][]string{}, "unreachable": unreachable}
	}
	distances := make(map[string]float64)
	parent := make(map[string]string)
	for node := range g.Nodes {
		distances[node] = math.Inf(1)
	}
	distances[start] = 0
	if !g.Weighted {
		visited := make(map[string]bool)
		queue := []string{start}
		visited[start] = true
		for len(queue) > 0 {
			current := queue[0]
			queue = queue[1:]
			neighbors := []string{}
			for neighbor := range g.AdjList[current] {
				neighbors = append(neighbors, neighbor)
			}
			sort.Strings(neighbors)
			for _, neighbor := range neighbors {
				if !visited[neighbor] {
					visited[neighbor] = true
					distances[neighbor] = distances[current] + 1
					parent[neighbor] = current
					queue = append(queue, neighbor)
				}
			}
		}
	} else {
		pq := make(priorityQueue, 0)
		heap.Init(&pq)
		heap.Push(&pq, &pqItem{node: start, priority: 0})
		for pq.Len() > 0 {
			item := heap.Pop(&pq).(*pqItem)
			current := item.node
			if item.priority > distances[current] {
				continue
			}
			for neighbor, weight := range g.AdjList[current] {
				newDist := distances[current] + weight
				if newDist < distances[neighbor] {
					distances[neighbor] = newDist
					parent[neighbor] = current
					heap.Push(&pq, &pqItem{node: neighbor, priority: newDist})
				}
			}
		}
	}
	unreachable := []string{}
	finalDistances := make(map[string]float64)
	finalPaths := make(map[string][]string)
	for node := range g.Nodes {
		if math.IsInf(distances[node], 1) {
			unreachable = append(unreachable, node)
		} else {
			finalDistances[node] = distances[node]
			path := []string{}
			for n := node; n != ""; n = parent[n] {
				path = append([]string{n}, path...)
				if n == start {
					break
				}
			}
			finalPaths[node] = path
		}
	}
	sort.Strings(unreachable)
	return map[string]interface{}{"distances": finalDistances, "paths": finalPaths, "unreachable": unreachable}
}

func (s *GraphStore) HasPath(graphID, start, end string) map[string]interface{} {
	s.mu.RLock()
	defer s.mu.RUnlock()
	g, ok := s.graphs[graphID]
	if !ok {
		return map[string]interface{}{"exists": false}
	}
	if start == end && g.Nodes[start] {
		return map[string]interface{}{"exists": true}
	}
	if !g.Nodes[start] || !g.Nodes[end] {
		return map[string]interface{}{"exists": false}
	}
	visited := make(map[string]bool)
	queue := []string{start}
	visited[start] = true
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		if current == end {
			return map[string]interface{}{"exists": true}
		}
		for neighbor := range g.AdjList[current] {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}
	return map[string]interface{}{"exists": false}
}

File: graphlib-1/ts/src/paths.ts
import { ShortestPathResult, AllShortestPathsResult, HasPathResult } from './types';
import { getGraph } from './graph';

class MinHeap {
  private heap: { node: string; distance: number }[] = [];

  push(node: string, distance: number): void {
    this.heap.push({ node, distance });
    this.bubbleUp(this.heap.length - 1);
  }

  pop(): { node: string; distance: number } | undefined {
    if (this.heap.length === 0) return undefined;
    const result = this.heap[0];
    const last = this.heap.pop()!;
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this.bubbleDown(0);
    }
    return result;
  }

  isEmpty(): boolean {
    return this.heap.length === 0;
  }

  private bubbleUp(index: number): void {
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      if (this.heap[parentIndex].distance <= this.heap[index].distance) break;
      [this.heap[parentIndex], this.heap[index]] = [this.heap[index], this.heap[parentIndex]];
      index = parentIndex;
    }
  }

  private bubbleDown(index: number): void {
    const length = this.heap.length;
    while (true) {
      const leftChild = 2 * index + 1;
      const rightChild = 2 * index + 2;
      let smallest = index;

      if (leftChild < length && this.heap[leftChild].distance < this.heap[smallest].distance) {
        smallest = leftChild;
      }
      if (rightChild < length && this.heap[rightChild].distance < this.heap[smallest].distance) {
        smallest = rightChild;
      }
      if (smallest === index) break;
      [this.heap[smallest], this.heap[index]] = [this.heap[index], this.heap[smallest]];
      index = smallest;
    }
  }
}

export function shortestPath(graphId: string, startNode: string, endNode: string): ShortestPathResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { exists: false, path: [], distance: -1 };
  }

  if (!graph.nodes.has(startNode) || !graph.nodes.has(endNode)) {
    return { exists: false, path: [], distance: -1 };
  }

  if (startNode === endNode) {
    return { exists: true, path: [startNode], distance: 0 };
  }

  if (graph.weighted) {
    return dijkstraPath(graph, startNode, endNode);
  } else {
    return bfsPath(graph, startNode, endNode);
  }
}

function bfsPath(graph: ReturnType<typeof getGraph>, startNode: string, endNode: string): ShortestPathResult {
  if (!graph) return { exists: false, path: [], distance: -1 };

  const visited = new Set<string>();
  const parent = new Map<string, string>();
  const queue: string[] = [startNode];
  visited.add(startNode);

  while (queue.length > 0) {
    const current = queue.shift()!;

    if (current === endNode) {
      const path: string[] = [];
      let node: string | undefined = endNode;
      while (node !== undefined) {
        path.unshift(node);
        node = parent.get(node);
      }
      return { exists: true, path, distance: path.length - 1 };
    }

    const neighbors = graph.adjacency.get(current);
    if (neighbors) {
      const sortedNeighbors = Array.from(neighbors.keys()).sort();
      for (const neighbor of sortedNeighbors) {
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          parent.set(neighbor, current);
          queue.push(neighbor);
        }
      }
    }
  }

  return { exists: false, path: [], distance: -1 };
}

function dijkstraPath(graph: ReturnType<typeof getGraph>, startNode: string, endNode: string): ShortestPathResult {
  if (!graph) return { exists: false, path: [], distance: -1 };

  const distances = new Map<string, number>();
  const parent = new Map<string, string>();
  const visited = new Set<string>();
  const heap = new MinHeap();

  for (const node of graph.nodes) {
    distances.set(node, Infinity);
  }
  distances.set(startNode, 0);
  heap.push(startNode, 0);

  while (!heap.isEmpty()) {
    const { node: current, distance: currentDist } = heap.pop()!;

    if (visited.has(current)) continue;
    visited.add(current);

    if (current === endNode) {
      const path: string[] = [];
      let node: string | undefined = endNode;
      while (node !== undefined) {
        path.unshift(node);
        node = parent.get(node);
      }
      return { exists: true, path, distance: currentDist };
    }

    const neighbors = graph.adjacency.get(current);
    if (neighbors) {
      for (const [neighbor, weight] of neighbors) {
        if (!visited.has(neighbor)) {
          const newDist = currentDist + weight;
          if (newDist < distances.get(neighbor)!) {
            distances.set(neighbor, newDist);
            parent.set(neighbor, current);
            heap.push(neighbor, newDist);
          }
        }
      }
    }
  }

  return { exists: false, path: [], distance: -1 };
}

export function allShortestPaths(graphId: string, startNode: string): AllShortestPathsResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { distances: {}, paths: {}, unreachable: [] };
  }

  if (!graph.nodes.has(startNode)) {
    return { distances: {}, paths: {}, unreachable: Array.from(graph.nodes) };
  }

  const distances: Record<string, number> = {};
  const paths: Record<string, string[]> = {};
  const parent = new Map<string, string>();
  const visited = new Set<string>();

  if (graph.weighted) {
    const heap = new MinHeap();
    const dist = new Map<string, number>();

    for (const node of graph.nodes) {
      dist.set(node, Infinity);
    }
    dist.set(startNode, 0);
    heap.push(startNode, 0);

    while (!heap.isEmpty()) {
      const { node: current, distance: currentDist } = heap.pop()!;

      if (visited.has(current)) continue;
      visited.add(current);
      distances[current] = currentDist;

      const neighbors = graph.adjacency.get(current);
      if (neighbors) {
        for (const [neighbor, weight] of neighbors) {
          if (!visited.has(neighbor)) {
            const newDist = currentDist + weight;
            if (newDist < dist.get(neighbor)!) {
              dist.set(neighbor, newDist);
              parent.set(neighbor, current);
              heap.push(neighbor, newDist);
            }
          }
        }
      }
    }
  } else {
    const queue: string[] = [startNode];
    const dist = new Map<string, number>();
    dist.set(startNode, 0);
    visited.add(startNode);
    distances[startNode] = 0;

    while (queue.length > 0) {
      const current = queue.shift()!;

      const neighbors = graph.adjacency.get(current);
      if (neighbors) {
        const sortedNeighbors = Array.from(neighbors.keys()).sort();
        for (const neighbor of sortedNeighbors) {
          if (!visited.has(neighbor)) {
            visited.add(neighbor);
            const newDist = dist.get(current)! + 1;
            dist.set(neighbor, newDist);
            distances[neighbor] = newDist;
            parent.set(neighbor, current);
            queue.push(neighbor);
          }
        }
      }
    }
  }

  // Build paths
  for (const node of visited) {
    const path: string[] = [];
    let current: string | undefined = node;
    while (current !== undefined) {
      path.unshift(current);
      current = parent.get(current);
    }
    paths[node] = path;
  }

  // Find unreachable nodes
  const unreachable: string[] = [];
  for (const node of graph.nodes) {
    if (!visited.has(node)) {
      unreachable.push(node);
    }
  }
  unreachable.sort();

  return { distances, paths, unreachable };
}

export function hasPath(graphId: string, startNode: string, endNode: string): HasPathResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { exists: false };
  }

  if (!graph.nodes.has(startNode) || !graph.nodes.has(endNode)) {
    return { exists: false };
  }

  if (startNode === endNode) {
    return { exists: true };
  }

  const visited = new Set<string>();
  const queue: string[] = [startNode];
  visited.add(startNode);

  while (queue.length > 0) {
    const current = queue.shift()!;

    const neighbors = graph.adjacency.get(current);
    if (neighbors) {
      for (const neighbor of neighbors.keys()) {
        if (neighbor === endNode) {
          return { exists: true };
        }
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          queue.push(neighbor);
        }
      }
    }
  }

  return { exists: false };
}
